Support for deframing in the application thread

ericgribkoff opened the pull request:
This pull request adds an option in `AbstractStream` (currently always disabled) to perform deframing in the application thread. The change is divided into three commits:

1. Introduces the `StreamListener.MessageProducer` interface and replaces `StreamListener.messageRead(InputStream)` with `StreamListener.messagesAvailable(MessageProducer)`.
2. Removes `MessageDeframer`'s tracking of the end-of-stream bit and checking for delivery stalled, and replaces them with a close request/callback mechanism. This is (arguably) simpler and better fits the model necessary for asynchronous client-thread deframing.
3. Adds the option for client-thread deframing, which places calls from the transport to the deframer inside of a `MessageProducer` that executes on the application thread. `MessageDeframer` remains entirely single-threaded, with the exception of a volatile boolean flag to signal an immediate close request.

This PR supersedes https://github.com/grpc/grpc-java/pull/2927 after offline discussion with @ejona86.

ejona86 commented on netty/src/test/java/io/grpc/netty/NettyClientTransportTest.java:
```diff
@@ -571,14 +573,16 @@ public void onReady() {
       this.method = method;
       this.headers = headers;
       stream.writeHeaders(new Metadata());
-      stream.request(1);
```
Why did this move?

ericgribkoff replied:
When deframing in the client thread (later commit, so this could be moved out of this commit entirely), calls to `request` trigger a call to `messagesAvailable` on the listener. Calling `request(1)` here in the `EchoServerStreamListener` constructor causes a null pointer exception, because the stream doesn't have a listener set until line 615, after the `EchoServerStreamListener` is created and set on the stream. This can be avoided by doing a null check in `AbstractStream`, but it seemed cleaner just not to invoke these methods when the listener isn't set.

But I realize now that this either needs to be documented on the `{Client,Server,}Stream` API and/or still do the a null check in `AbstractStream.messagesAvailable`

ericgribkoff replied:
Done.

ejona86 commented on netty/src/test/java/io/grpc/netty/NettyStreamTestBase.java:
```diff
@@ -129,11 +133,12 @@ public void inboundMessageShouldCallListener() throws Exception {
-    ArgumentCaptor<InputStream> captor = ArgumentCaptor.forClass(InputStream.class);
-    verify(listener()).messageRead(captor.capture());
+
+    InputStream message = listenerMessageQueue.poll();
```
This being populated by the extending class is really subtle. I think we've instead made an abstract method here (say, `listenerMessageQueue()`) that is required to be implemented by extending classes.

ericgribkoff replied:
Done

ejona86 commented on testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java:
```diff
@@ -177,6 +178,22 @@ public void setUp() {
         .thenReturn(serverStreamTracer);
     callOptions = CallOptions.DEFAULT.withStreamTracerFactory(clientStreamTracerFactory);
+
+    doAnswer(
```
Eww... each test needing to do something different is pretty ugly. It could be made better if you made the Answers non-anonymous. But I do wonder if there is something better we can do. Maybe something with Captors... Dunno.

ericgribkoff replied:
Each test can do the same thing - store the messages it receives - but a few didn't need this, so I just loop over them (and don't close them, but correcting that). This can definitely be made nicer (using a concrete class instead of mocks, as you suggested in AbstractServerStreamTest), but I didn't want to invest too much additional time into redoing the tests until the API changes are approved/stabilized, as changes there could require completely different testing mechanism.

Incidentally, this change includes a couple fixes to the flow control pushback test that - in local testing - seem to deflake it. That's orthogonal to this, but hopefully a nice benefit once I invest the time to clean up the mocking/Answers used here.

ejona86 commented on testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java:
```diff
@@ -523,6 +540,18 @@ public void newStream_duringShutdown() throws Exception {
+            public Void answer(InvocationOnMock invocation) throws Throwable {
+              StreamListener.MessageProducer producer =
+                  (StreamListener.MessageProducer) invocation.getArguments()[0];
+              while (producer.next() != null) {}
```
I'd really hope we'd close the messages properly everywhere in AbstractTransportTest.

ericgribkoff replied:
Done

ericgribkoff commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -466,22 +466,25 @@ public final void runInContext() {
+              }
+              try {
+                observer.onMessage(method.parseResponse(message));
+              } finally {
```
Can you elaborate on what you're looking for here? This can fail to close some messages, if `producer.next()` throws an exception, but not sure what we should do about that here since the messages might not exist or have been deframed at that point. When the producer is a wrapper around a single message, as in this commit, isn't the try/finally here exactly the same as with the current code?

ericgribkoff replied:
Fixed. We spoke offline and the problem would occur with exceptions in handling the message when the producer potentially (as in the subsequent parts of this PR) has additional queued messages.

ericgribkoff commented on core/src/test/java/io/grpc/internal/ClientCallImplTest.java:
```diff
@@ -820,5 +823,21 @@ void release() {
     }
   }
+
+  private static class SingleMessageProducer implements StreamListener.MessageProducer {
```
Was waiting to make sure this was the right approach. Done.

ericgribkoff commented on core/src/test/java/io/grpc/internal/AbstractServerStreamTest.java:
```diff
@@ -68,6 +74,22 @@ public WritableBuffer allocate(int capacityHint) {
   @Test
   public void frameShouldBeIgnoredAfterDeframerClosed() {
     ServerStreamListener streamListener = mock(ServerStreamListener.class);
```
Done. This would be a nicer approach in the other tests as well, but I'd prefer to hold off on that as a follow-up PR, after everything else is solidified.

ericgribkoff commented on core/src/main/java/io/grpc/internal/ServerCallImpl.java:
```diff
@@ -237,26 +236,27 @@ public void cancelled(Context context) {
+          }
+          try {
+            listener.onMessage(call.method.parseRequest(message));
+          } finally {
```
Fixed

ericgribkoff commented on core/src/main/java/io/grpc/internal/ServerImpl.java:
```diff
@@ -497,11 +497,14 @@ public LogId getLogId() {
+        try {
+          message.close();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
```
Fixed

ejona86 commented on core/src/main/java/io/grpc/internal/AbstractStream.java:
```diff
@@ -162,30 +162,28 @@ public void messageRead(InputStream is) {
      * Closes this deframer and frees any resources. After this method is called, additional calls
```
This documentation _really_ needs to be updated to note that closing may be queued behind deframing messages.

ericgribkoff replied:
This is done in the third commit. If this is all eventually to be merged as three separate commits, I can redo the history later to get the comment in the second commit instead.

ejona86 commented on core/src/main/java/io/grpc/internal/AbstractServerStream.java:
```diff
@@ -232,9 +244,21 @@ public void inboundDataReceived(ReadableBuffer frame, boolean endOfStream) {
+                closeListener(status);
+              }
+            };
+        closeDeframer(false);
```
Previously the code notified the application immediately, right? Why isn't this `true`?

ericgribkoff replied:
Done, and added test case to AbstractServerStreamTest.

ejona86 commented on core/src/main/java/io/grpc/internal/AbstractServerStream.java:
```diff
@@ -243,7 +267,19 @@ public final void transportReportStatus(Status status) {
+                closeListener(Status.OK);
+              }
+            };
+        closeDeframer(false);
```
This seems like it should be `true` as well to match the previous code. Generally, on the server-side we don't continue deframing after we have a status. This seems like it could cause a deadlock (in that we make no further progress). I'd have sort of expected a test to catch this... maybe we're missing a test case here.

ericgribkoff replied:
Done, and added test case to AbstractServerStreamTest. It seemed to fit better there than in the abstract transport test, as it's fairly low level.

ejona86 commented:
I got through the second part. I still need to look at the third.

ericgribkoff commented on core/src/main/java/io/grpc/internal/MessageDeframer.java:
```diff
@@ -253,26 +251,8 @@ private void deliver() {
       boolean stalled = unprocessed.readableBytes() == 0;
-
-      if (endOfStream && stalled) {
-        boolean havePartialMessage = nextFrame != null && nextFrame.readableBytes() > 0;
```
Added this back in, and added test cases in AbstractServerStreamTest and MessageDeframerTest. The deframer will notify the listener on close if it has a partial message - this is ignored by clients, since they don't set end-of-stream, but used by server to call deframeFailed when appropriate.

ejona86 replied:
It being missed on clients is an oversight (probably lost in an earlier refactoring). I've filed https://github.com/grpc/grpc-java/issues/3264

ericgribkoff commented:
@ejona86 I squashed the updates to the first two commits and did a force-push to clean up the PR. As we discussed, I also updated the third commit (adding app-thread deframing support) to use ApplicationThreadDeframer, a new class which handles the interactions between the transport state and MessageDeframer.

ejona86 commented on okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java:
```diff
@@ -941,7 +943,8 @@ public void headers(boolean outFinished,
     public void rstStream(int streamId, ErrorCode errorCode) {
-      finishStream(streamId, toGrpcStatus(errorCode).augmentDescription("Rst Stream"), null, null);
+      finishStream(
+          streamId, toGrpcStatus(errorCode).augmentDescription("Rst Stream"), false, null, null);
```
This needs logic to determine whether to stopDelivery.

ericgribkoff replied:
Done. I restored the original check for CANCELLED or DEADLINE_EXCEEDED status alongside the explicit stopDelivery flag.

ericgribkoff commented on core/src/main/java/io/grpc/internal/ApplicationThreadDeframer.java:
```diff
@@ -0,0 +1,189 @@
+      TransportExecutor transportExecutor) {
+    this.deframer =
+        new MessageDeframer(this, decompressor, maxMessageSize, statsTraceCtx, debugString);
+    this.storedListener = listener;
```
Done

ericgribkoff commented on core/src/main/java/io/grpc/internal/ApplicationThreadDeframer.java:
```diff
@@ -0,0 +1,189 @@
+
+  @Override
+  public void messagesAvailable(StreamListener.MessageProducer producer) {
+    messageReadQueue.add(producer.next());
```
Done

ericgribkoff commented on core/src/main/java/io/grpc/internal/MessageDeframer.java:
```diff
@@ -105,29 +116,20 @@ public MessageDeframer(Listener listener, Decompressor decompressor, int maxMess
-   */
+  @Override
   public void setDecompressor(Decompressor decompressor) {
+    if (isClosedOrScheduledToClose()) {
```
You're right. Removed.

ericgribkoff commented on core/src/main/java/io/grpc/internal/AbstractClientStream.java:
```diff
@@ -174,7 +174,7 @@ public final boolean isReady() {
 
     private boolean deframerClosed = false;
-    private Runnable deframerClosedTask;
+    protected Runnable deframerClosedTask;
```
Did this slip in from an earlier review? You're right, but I don't think the commit where this was protected survives anymore - it was removed in the earlier rebase, and so this should no longer be an issue.

ejona86 replied:
Yeah, that looks like a comment that failed to be sent out earlier. Sorry for the noise.

ejona86 commented on okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java:
```diff
@@ -721,8 +732,14 @@ void finishStream(
+              .transportState()
+              .transportReportStatus(
+                  status,
+                  isCancelled || stopDelivery,
```
This is complicated and surprising. I had hoped only the RST stream method I mentioned earlier would have this logic.

ericgribkoff replied:
The stop on CANCELLED or DEADLINE_EXCEEDED is the existing logic here. I should not have removed this entirely in part 3 of this PR - I now am just adding it back to preserve the current behavior, alongside the now-necessary explicit stopDelivery flag. 

Do you want this logic moved to the rst stream method? I was trying to minimize changes to other code paths, which would also potentially be impacted by moving the isCancelled check - this might be the appropriate place for the check, but didn't seem relevant to the changes here.

ejona86 replied:
I'm not excited about the mixing of implicit (via Status) and explicit `stopDelivery`. It should be one way or the other, unless there was a _really good_ reason. In auditing the existing calls to this method, it seemed that only the rst stream one was depending on this behavior. If you'd like we can sit down and pair program/audit the callers.

ejona86 commented on core/src/main/java/io/grpc/internal/ServerCallImpl.java:
```diff
@@ -237,27 +241,37 @@ public void cancelled(Context context) {
-          return;
+        while ((message = producer.next()) != null) {
+          if (call.cancelled) {
+            return;
```
Getting here fails to close `message`.

ejona86 commented on core/src/main/java/io/grpc/internal/ClientCallImpl.java:
```diff
@@ -457,24 +458,36 @@ public final void runInContext() {
-              message.close();
+            while ((message = producer.next()) != null) {
+              if (closed) {
+                return;
```
Getting here fails to close `message`.

ejona86 commented on okhttp/src/main/java/io/grpc/okhttp/OkHttpClientStream.java:
```diff
@@ -150,7 +150,7 @@ public void request(final int numMessages) {
     public void cancel(Status reason) {
       synchronized (state.lock) {
-        state.cancel(reason, false, null);
+        state.cancel(reason, true, null);
```
Yeah, always stopDelivery. If it's not today, it should be.

ericgribkoff commented:
@ejona86 Just updated the handling of exceptions/closing in ClientCallImpl and ServerCallImpl. Should be somewhat improved now in terms of clarity, and maintains for both server and client that (a) all messages will be closed (b) if the application code throws an exception, that exception is preserved (e.g., an IOException during the clean-up phase's closes is ignored) and (c) if a message close fails after the application message handling returned normally, that IOException is propagated just like in the current code.

ericgribkoff commented:
Rebased onto master to fix the merge conflicts and squashed the fixup commits back into the original three commit structure.
